test/mockserver/mockspanner.ts

```diff
@@ -0,0 +1,730 @@
+   */
+  static create(): MockSpanner {
+    return new MockSpanner();
+  }
+
+  /**
+   * Registers a result for an SQL statement on the server.
+   * @param sql The SQL statement that should return the result.
+   * @param result The result to return.
+   */
+  putStatementResult(sql: string, result: StatementResult) {
+    this.statementResults.set(sql, result);
+  }
+
+  removeExecutionTime(fn: Function) {
+    this.executionTimes.delete(fn.name);
+  }
+
+  setExecutionTime(fn: Function, time: SimulatedExecutionTime) {
+    this.executionTimes.set(fn.name, time);
+  }
+
+  abortTransaction(transaction: Transaction): void {
+    const formattedId = `${transaction.session.formattedName_}/transactions/${transaction.id}`;
+    if (this.transactions.has(formattedId)) {
+      this.transactions.delete(formattedId);
+      this.transactionOptions.delete(formattedId);
+      this.abortedTransactions.add(formattedId);
+    } else {
+      throw new Error(`Transaction ${formattedId} does not exist`);
+    }
+  }
+
+  freeze() {
+    this.frozen++;
+  }
+
+  unfreeze() {
+    if (this.frozen === 0) {
+      throw new Error('This mock server is already unfrozen');
+    }
+    this.frozen--;
+  }
+
+  /**
+   * Creates a new session for the given database and adds it to the map of sessions of this server.
+   * @param database The database to create the session for.
+   */
+  private newSession(database: string): protobuf.Session {
+    const id = this.sessionCounter++;
+    const name = `${database}/sessions/${id}`;
+    const session = protobuf.Session.create({name, createTime: now()});
+    this.sessions.set(name, session);
+    return session;
+  }
+
+  private static createSessionNotFoundError(name: string): grpc.ServiceError {
+    const error = new Error(`Session not found: ${name}`);
+    return Object.assign(error, {
+      code: grpc.status.NOT_FOUND,
+    });
+  }
+
+  private static createTransactionNotFoundError(
+    name: string
+  ): grpc.ServiceError {
+    const error = new Error(`Transaction not found: ${name}`);
+    return Object.assign(error, {
+      code: grpc.status.NOT_FOUND,
+    });
+  }
+
+  private static createTransactionAbortedError(
+    name: string
+  ): grpc.ServiceError {
+    const error = new Error(`Transaction aborted: ${name}`);
+    const metadata = new grpc.Metadata();
+    const retry = RetryInfo.encode({
+      retryDelay: {
+        seconds: 0,
+        nanos: 1,
+      },
+    });
+    metadata.add(RETRY_INFO, Buffer.from(retry.finish()));
+    return Object.assign(error, {
+      code: grpc.status.ABORTED,
+      metadata,
+    });
+  }
+
+  private static sleep(ms): Promise<void> {
+    return new Promise(resolve => setTimeout(resolve, ms));
+  }
+
+  private async simulateExecutionTime(functionName: string): Promise<void> {
+    while (this.frozen > 0) {
+      await MockSpanner.sleep(10);
+    }
+    const execTime = this.executionTimes.get(functionName);
+    if (execTime) {
+      if (
```

Comment:
What do you think about merging this `if` block with previous one? Something like:

```js
if (execTime && execTime.errors && execTime.errors.length && !execTime.errors[0].streamIndex) {
 // ...
}
```

Reply:
I agree, that is better.